another try to fix intermittent macOS packaging failures

imachug opened the pull request:
Resolves: bugs in #198

- Try to detach the image several times during normal build (as in #211)
- Move the script to the correct directory (because it's not only for CI)
- Add 'set -e' for safety (it's for error checking @emabrey)
- Handle spaces in filenames correctly (because `$*` doesn't handle spaces and `""$*""` doesn't do anything at all)
- Use correct path to repeat_hdiutil.sh in CI (i.e. an absolute path. The code is python-version agnostic)

- [x] I have signed off my commits using `-s` or `Signed-off-by`\* (See: [Contributing § DCO](https://github.com/tenacityteam/tenacity/blob/master/CONTRIBUTING.md#developer-certificate-of-origin))
- [ ] I made sure the code compiles on my machine [I now have access to mac so I checked the script but I haven't try building with cmake]
- [x] I made sure there are no unnecessary changes in the code\*
- [x] I made sure the title of the PR reflects the core meaning of the issue you are solving\*
- [x] I made sure the commit message(s) contain a description and answer the question "Why do those changes fix that particular issue?" or "Why are those changes really necessary as improvements?"\*

\* indicates required

caughtquick commented:
Until isn't really used and may hurt usability

falkTX commented on scripts/ci/package.sh:
```diff
@@ -9,7 +9,7 @@ cd build
     cmake --build . --target innosetup --config "${AUDACITY_BUILD_TYPE}"
 else
-    export CPACK_COMMAND_HDIUTIL="./macos/repeat_hdiutil.sh"
+    export CPACK_COMMAND_HDIUTIL="$(python -c "import os, sys; print(os.path.dirname(os.path.realpath(sys.argv[1])))" "$0")/../../mac/scripts/repeat_hdiutil.sh"
```
this is awful...

imachug replied:
This is not awful, this is the easiest way to get something like `realpath` using built-in tools. And I must say it's quite readable too.

falkTX replied:
If you read the rest of the script, it uses local paths. So using "." here is fine too. No need to call into python here ouch

imachug replied:
Just for the record (although we've discussed this on IRC): I have absolutely no idea whether CPACK_COMMAND_HDIUTIL supports a relative path or not, because it's handled by CPack somewhere internally, and it will break if CPack does chdir(), and we haven't even managed to get it working yet so I'm playing safe and using an absolute path.

falkTX commented on mac/scripts/repeat_hdiutil.sh:
```diff
@@ -0,0 +1,19 @@
+counter=0
+num_secs_await_retry=1
+
+while ! /usr/bin/hdiutil "$@"; do
```
the current syntax was fine. just because you dont understand what `until` does, doesn't make it right to change it.

imachug replied:
Just because you think I'm dumber than you doesn't mean I am dumber than you. I do know what `until` does, but I also haven't seen `until` during reading others' bash scripts for 5 years. No modern language has `until`, and `while !` is quite a popular alternative.

falkTX replied:
No need to feel offended. I also didn't know that `until` was a thing in bash until seeing this.
But it makes perfect sense, and now I know something new. win-win.

imachug replied:
Want to feel different and use `until`? Go for it, I don't care as long as the code works. I just used a much more popular alternative, but if you just want `until` for no reason go for it.

emabrey replied:
The reason I used `until` is because in general I try to avoid extra logic when possible. `until` and `while !` are entirely identical programmatically, it's just a matter of which syntax you prefer. It's a preference thing, not really a problem with my original commit. It's kind of like `while` being syntactic sugar for `goto`.

falkTX commented:
You know this isn't nice right? Why do you go over what another person did?
Leaving inline comments on the previous PR is possible, no need to step over anyone's work.
Please don't do that again.

imachug commented:
You didn't even fucking bother to read changelog and now you're saying I'm going over what another person did? That's called bugfixing.

caughtquick commented:
> You didn't even fucking bother to read changelog and now you're saying I'm going over what another person did? That's called bugfixing.

Can you actually explain what this does better than that?

imachug commented:
1. I replaced `hdiutil` with `"${SRCROOT}/scripts/repeat_hdiutil.sh" detach "${ATTACHED}"` in `mac/scripts/build_dist.sh`. This makes sure the disk gets detached correctly not only when CI is run but when `build_dist.sh` is executed.
2. As a consequence of the above, I moved the script to `mac/scripts` and also deleted the auxiliary `ci/scripts/macos` directory because the script is used not only during CI.
3. I also returned `set -e` back. What `set -e` does is make bash exit with an error code on the first failure. Removing `set -e` just hides all errors that occur during script execution. I think it was missing from the original PR because the script failed with `set -e`, i.e. it crashed because of bugs.
4. I fixed handling files with spaces. Say, let's consider a script that's run like this: `./test.sh hello "world aba"`. In this case `$*` would mean `hello world aba`, i.e. the second argument would be split into two arguments. `""$*""` (as used in the original script) is just `""hello world aba""` which again gets unquoted into `hello world aba`, the same problem. `"$@"` on the other hand correctly handles spaces and yields `hello "world aba"`. In this sense, `$@` is not just a variable, `"$@"` is a special syntax on itself.
5. I also used an absolute path to repeat_hdiutil.sh in CI. A relative path won't work if CPack changes working directory; I'm not sure if it actually does so, and the docs about this CPack option on the internet don't say whether relative path is supported, and all examples I could find on the internet use absolute path, so I used an absolute path just in case.

falkTX commented:
That is very nice, but redoing the PR doesnt seem like the way to go.
Specially calling it invalid code.

I propose the modify the file with the changes you have in mind first, then take care of the rest.

caughtquick commented:
I'm reopening this so we can have a productive conversation. Please refrain from saying phrases like "invalid code".

imachug commented:
> I propose the modify the file with the changes you have in mind first, then take care of the rest.

I actually did exactly that: I renamed the file and applied one or two fixes. It's just that Git doesn't recognize that as file renaming.

falkTX commented:
Also, please don't call into python during the build script.
The "." relative usage is fine, specially considering other parts of the script use that too.

Semisol commented:
I am going to step in.
I request that you all calm down and stop throwing insults at each other, that gets us nowhere.

caughtquick commented:
Yep I agree, adding Python seems like unnecessary complication, it if can be done in bash that would be better.

imachug commented:
> Also, please don't call into python during the build script.

No problem, that looks like a hack to me too, but I don't know an alternative unfortunately.

> The "." relative usage is fine, specially considering other parts of the script use that too.

There is difference between using `.` *inside* script and passing it in a variable to another program.

caughtquick commented:
The MacOS build just failed @imachug

imachug commented:
Strange, it looks like our script wasn't called at all?

emabrey commented:
Don't add -e. It will break the script because it fails fast, meaning the first time `hdiutil` doesn't work it kills the retry script so  it won't retry.

emabrey commented on mac/scripts/build_dist.sh:
```diff
@@ -205,7 +205,7 @@ cp "${SRCROOT}/Resources/Audacity-DMG-background.png" "${DMG}/.background"
 ATTACHED=$(hdiutil info | awk "/\/Volumes\/${VOL}/{print \$1}")
 if [ -n "${ATTACHED}" ]
 then
-   hdiutil detach "${ATTACHED}"
```
Don't replace this. The `hdiutil` package does not have a bug- CPack does. The only reason that you'd replace this usage of `hdiutil` is if that package was bugged, which I don't believe it is (though maybe I'm wrong?).

imachug replied:
What I am afraid of is if someone or something mounts the volume, like Finder for example. I'm not sure if that applies here or not; I've had very similar problems under *nix but maybe the situation here is different. I can rollback this change if you confirm this change is useless.

emabrey replied:
So far the only place we have had an issue is inside the CPack usage of `hdiutil` so I don't think we should mess with this until/if it breaks. As far as I've seen, `build_dist.sh` doesn't fail.

imachug replied:
Okay, agreed, I will rollback this then.

emabrey commented on mac/scripts/repeat_hdiutil.sh:
```diff
@@ -0,0 +1,19 @@
+
+((${BASH_VERSION%%.*} >= 4)) || { echo >&2 "$0: Error: Please upgrade Bash."; exit 1; }
+
+set -euxo pipefail
```
Using -e is a good idea (no I did not disable it because the script was buggy, I have no idea why you said that). If we use it then we HAVE to keep the call to hdiutil inside a `while` or an `until` block (specifically the first line).

imachug replied:
> I have no idea why you said that

I got that from the commit message. Sorry for misunderstanding.

> If we use it then we HAVE to keep the call to hdiutil inside a while or an until block (specifically the first line).

Wait a second. What we want is to repeat `hdiutil` until it returns success, right? And `while ! ...` does exactly that, doesn't it?

emabrey replied:
Yes. My point is that if someone else comes along and changes it say a `for` loop or uses anything but `while` or `until` it will break. `set -e` doesn't apply to the first line of a `while` loop or an `until` loop.

imachug replied:
I see what you mean but I think I can't agree with you. I think that, in general, any command with unhandled non-zero exit code may be a potential bug in the code. I don't think anyone would reasonably do something like

```bash
while true; do
    hdiutil "$@"
    ...
done
```

because `hdiutil` may fail, and it is thus necessary to check its exit code somehow, using `if` or whatever. I'm not sure if shellcheck checks this but I think this would be a reasonable diagnostic.

emabrey commented on scripts/ci/package.sh:
```diff
@@ -9,8 +9,7 @@ cd build
 else
-    export CPACK_COMMAND_HDIUTIL="./macos/repeat_hdiutil.sh"
-    cpack -C "${AUDACITY_BUILD_TYPE}" --verbose
+    cpack -C "${AUDACITY_BUILD_TYPE}" --verbose -DCPACK_COMMAND_HDIUTIL="$(python -c "import os, sys; print(os.path.dirname(os.path.realpath(sys.argv[1])))" "$0")/../../mac/scripts/repeat_hdiutil.sh"
```
-D doesn't work here. We don't want to configure CPack, we want a bash environment variable. The cpack man page says that `-D` is used to "Set a CPack variable [...] that can be used by the generator." Well the whole point of the call to `hdiutil` is we aren't in a CPack generator anymore. That is why your script didn't get called.

imachug replied:
`CPACK_COMMAND_HDIUTIL` doesn't work either, we've checked that.

emabrey commented:
Btw using `""$*""` was because I wanted the command line args of the script added onto the hdiutil command without ANY editing. It is possible that it will not work, but as far as I can tell it would be because the caller of the script didn't properly format their call. How does using `""$*""` mangle the input to the script args?

imachug commented:
> Btw using ""$*"" was because I wanted the command line args of the script added onto the hdiutil command without ANY editing. It is possible that it will not work, but as far as I can tell it would be because the caller of the script didn't properly format their call. How does using ""$*"" mangle the input to the script args?

Suppose I call the script like `./repeat_hdiutil.sh 1 "2 3"`. `""$*""` in this context would evaluate to `""[first argument] [second argument]""`, i.e. `""1 2 3""`. `""1` and `3""` are interpreted as string concatenation so we basically have three arguments now: 1, 2 and 3.

Maybe you meant `"$*"` with a single quote sign but that would be parsed as `"1 2 3"` (i.e. a single argument) too.

It does not really *mangle* the input. It was more of a design failure when people were designing bash. `"$*"` is parsed as `"$1 $2 ..."`. What you really want is `"$1" "$2" ...`, and that's exactly what `"$@"` does.

imachug commented:
> Also it isn't proper to mix $@, array syntax, with strings (usually). Shellcheck gives a warning about this.

`$@` is not array syntax. Neither is `"$@"`. `${var[@]}` *would* be array syntax but that's a different matter and we aren't using this.

emabrey commented:
> 
> 
> > Also it isn't proper to mix $@, array syntax, with strings (usually). Shellcheck gives a warning about this.
> 
> `$@` is not array syntax. Neither is `"$@"`. `${var[@]}` _would_ be array syntax but that's a different matter and we aren't using this.

https://tiswww.case.edu/php/chet/bash/bashref.html#Special-Parameters

So if we use your solution and they have already string escaped, won't it actually do it again?

`repeat_hdiutil.sh "test" "test2" "test3"` results in  `repeat_hdiutil.sh ""test"" ""test2"" ""test3""` with your method, does it not?

Be-ing commented:
Please use more descriptive titles for your pull requests.

imachug commented:
> Using repeat_hdiutil.sh "$@" results in repeat_hdiutil.sh ""test"" ""test2"" ""test3"" with your method, does it not?

No, it's actually the other way round... I think I see where the misunderstanding comes from so let me explain.

Reiterating the obvious: let's say you open your favorite shell, type `./script 1 "2 3" 4` and press enter. The shell parses your command. First it sees `./script` followed by a space, which goes into `argv[0]`. Then it sees `1` followed by a space, which goes into `argv[1]`. Then it sees a quotation sign so it parses the string till the nearest quotation sign and `2 3` goes into `argv[2]`. Then `4` goes into `argv[3]`. Speaking in C++-like language, you have `argv = {"./script", "1", "2 3", "4"}` at the moment. See, the quotation signs are not present here *and* the strings and escapes are parsed by the caller, not the callee.

Then the script is interpreted with bash. You probably know that if the bash script says something like `rm $2`, bash substitutes `argv[2]` verbatim, i.e. executes `rm 2 3`. This wouldn't delete file `2 3` but files `2` and `3` so to workaround that you use quotes: `rm "$2"`.

Now pretend that you want to delete not only file in `argv[2]` but also `argv[1]` and `argv[3]` and all other arguments. You want something like: `rm "$1" "$2" "$3" ...`.

Back in 20th century you had a metavariable `$*`, which means the same as `$1 $2 $3 ...`. So if you did `rm $*` it would delete 4 files: `1`, `2`, `3` and `4`, which is not what you want. You could try to add quotes: `rm "$*"` but that is the same as `rm "$1 $2 $3 $4"` so that would delete file named `1 2 3 4`.

So some bash dev figured out there's no reliable way to handle multiple arguments with spaces so they invented `$@`. Here's what it does: if you use it as-is, like `rm $@`, it works just like `$*`, nothing strange here. But when it's spelled like `"$@"` bash replaces this with `"$1" "$2" "$3" ...`, so `rm "$@"` would delete files `1`, `2 3` and `4` as expected.

This is all very different in Windows where the argument string is parsed by the callee, not the caller. Batch, for instance, is much more stupid than bash and almost works on string level, not token level. In Batch scripts, for instance ,`%n` *includes* quotes if they are present, so `del %2` would be interpreted as `del "2 3"` and would delete the correct file; similarly `del %*` would mean `del %1 %2 %3 ...` and would delete files `1`, `2 3` and `4` as expected. So in Batch, `%*` kinda matches `"$@"`. And, by the way, if you wanted the behavior similar to `$*`, you can't: there's `%~n` which means argument `%n` without quotes but there is nothing like `%~*`.

Hope this answers your question.

emabrey commented:
This is what each of them do (ignore the `set +x` obviously):

```
$@:
+ echo test test2 test3 test4
test test2 test3 test4
+ set +x
$*:
+ echo test test2 test3 test4
test test2 test3 test4
+ set +x
"$@":
+ echo test 'test2 test3' test4
test test2 test3 test4
+ set +x
"$*":
+ echo 'test test2 test3 test4'
test test2 test3 test4
+ set +x
""$*"":
+ echo test test2 test3 test4
test test2 test3 test4
+ set +x
""$*"":
+ echo test test2 test3 test4
test test2 test3 test4
+ set +x
```

So we do want "$@", so that's a good catch, but you are incorrect about what it does in detail. `$*` processes the argument list as a concatenated string. `$@` process it as an array of arguments (which is why it shares the syntax of using `@` like I mentioned earlier). `$@` actually aliases to each of the positional parameters (`$0`, `$1`, `$2`) which is exactly what `@` does in other array based contexts. I didn't realize that `$*` strips the quotes off before concatenation, so you are right to use `$@`. Since we want to keep the quotes, we need to use `"$@"` like you said. I incorrectly assumed that because `$*` was a concatenated string it would just give me the raw input that I wanted (placing the burden of stringifying space-containing args on the caller).

imachug commented:
> but you are incorrect about what it does in detail.

That's because you're using `echo` for demonstration. Try using something like `printf "%s\n" "$@"` instead. This would print every argument on its own line. `printf "%s\n" $*` would split every argument by spaces too, and so on.
